span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java

```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
     };
   }
 
-  private void sanitiseSpan(Builder rawSpanBuilder) {
-
+  // redact PII tags, tag comparisons are case-insensitive (Resource tags are skipped)
+  private void redactSpan(Builder rawSpanBuilder) {
     try {
       var attributeMap = rawSpanBuilder.getEvent().getAttributes().getAttributeMap();
-      var spanServiceName = rawSpanBuilder.getEvent().getServiceName();
+
+      Integer piiFieldsCount = 0;
+      Integer pciFieldsCount = 0;
+
       Set<String> tagKeys = attributeMap.keySet();
+      for (PIIPCIField piiPciField : PIIPCIFields) {
+        Matcher matcher = null;
+        if (piiPciField.getRegexInfo().isPresent()) {
+          matcher = piiPciField.getRegexInfo().get().getRegexPattern().matcher("");
+        }
+        PIIPCIFieldType piiPciFieldType = piiPciField.getPiiPciFieldType();
+        for (String tagKey : tagKeys) {
+          /* A Simple regex like PAN no. can have false positives.
+             Hence, we only redact tag val when tagKey is present in possible list of regexTagKeySet.
+             If regexTagKeySet is empty, we redact the tag val always.
+          */
+          if (!piiPciField.getTagKeySet().isEmpty()
+              && !piiPciField.getTagKeySet().contains(tagKey)) {
+            continue;
+          }
 
-      AtomicReference<Boolean> containsPIIFields = new AtomicReference<>();
-      containsPIIFields.set(false);
-
-      try {
-        tagKeys.stream()
-            .filter(
-                tagKey ->
-                    tagKeysToRedact.contains(tagKey.toUpperCase())
-                        || attributeMap
-                            .get(tagKey)
-                            .getValue()
-                            .equals(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL))
-            .peek(
-                tagKey -> {
-                  containsPIIFields.set(true);
-                  spanAttributesRedactedCounters
-                      .computeIfAbsent(
-                          PIIMatchType.KEY.toString(),
-                          k ->
-                              registerCounter(
-                                  SPAN_REDACTED_ATTRIBUTES_COUNTER,
-                                  Map.of("matchType", PIIMatchType.KEY.toString())))
-                      .increment();
-                })
-            .forEach(
-                tagKey -> {
-                  logSpanRedaction(tagKey, spanServiceName, PIIMatchType.KEY);
-                  attributeMap.put(tagKey, redactedAttributeValue);
-                });
-      } catch (Exception e) {
-        LOG.error("An exception occurred while sanitising spans with key match: ", e);
-      }
+          boolean containsSensitiveData = false;
+          if (piiPciField.getRegexInfo().isPresent()) {
+            assert matcher != null;
+            matcher.reset(attributeMap.get(tagKey).getValue());
```

Comment:
Why do you need a reset when you're creating a new matcher object for each tag key-val basically?